client/state/data-layer/wpcom-http/utils.js

```diff
@@ -158,3 +163,73 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options ) => ( s
 
 	return initiator( store, action );
 };
+
+/*
+ * Converts an application-level Calypso action that's handled by the data-layer middleware
+ * into a low-level action. For example, HTTP request that's being initiated, or a response
+ * action with a `meta.dataLayer` property.
+ */
+const createRequestAction = ( options, action ) => {
+	const {
+		initiate = noop,
+		onSuccess = noop,
+		onError = noop,
+		onProgress = noop,
+		fromApi = identity,
+	} = options;
+
+	const data = getData( action );
+	if ( data ) {
+		try {
+			return onSuccess( action, fromApi( data ) );
```

Comment:
Does this mean users of this code should throw an exception on invalid data rather than dispatching `onError` like we do now?

Reply:
This `try/catch` block merely catches errors thrown by the `fromApi` transformer. Even if the network request is successful and the `onSuccess` handler is called, the final result can still be error.

If the network request fails, calling `onError` is still the right thing. There is no change from `dispatchRequest` here.